test/Test.hs

```diff
@@ -1,7 +1,191 @@
+{-# language ScopedTypeVariables #-}
+{-# language DeriveGeneric #-}
+{-# language FlexibleContexts #-}
+{-# language UndecidableInstances #-}
+{-# language GeneralizedNewtypeDeriving #-}
+{-# language DeriveFunctor #-}
+{-# language StandaloneDeriving #-}
 module Main(main) where
 
-import Test.Hspec
+import Control.Monad.IO.Class (liftIO)
+import Hedgehog (MonadGen, Range)
+import qualified Hedgehog.Gen as Gen
+import Hedgehog.Range (Size)
+import qualified Hedgehog.Range as Range
+import Test.Hspec (before, describe, hspec, it, shouldBe)
+import Test.Hspec.Hedgehog (PropertyT, diff, forAll, hedgehog, (/==), (===))
+import Control.Monad.Logic.Sequence
+import Control.Monad.Logic.Sequence.Compat
+import Control.Monad.Logic.Sequence.Internal (SeqT (..))
+import Data.SequenceClass hiding ((:<), empty)
+import qualified Data.SequenceClass as S
+import Control.Monad.Logic.Sequence.Internal.Queue
+import Data.Functor.Identity
+import Control.Applicative
+import Data.Function (fix)
+import GHC.Generics (Generic)
+import qualified Hedgehog.Function as Fun
+import Data.Foldable (foldl', for_)
+import Control.Monad.Free (Free (..))
+import Data.Functor.Classes
+import qualified Control.Monad.Logic as L
+import Debug.Trace (trace)
+import Control.Monad.Trans.Maybe
+import Control.Monad.Reader
+import Control.Monad.Except
+import Control.Monad.Morph (hoist)
+
+-- | A generic "container" functor. We can use this with `Free` to get
+-- an inspectable `Monad` that's unlikely to hide any mistakes we make.
+data TestF a = TestF !Int [a]
+  deriving (Show, Eq, Generic, Functor)
+
+instance Eq1 TestF where
+  liftEq eq (TestF i1 xs1) (TestF i2 xs2)
+    = i1 == i2 && liftEq eq xs1 xs2
+
+instance Show1 TestF where
+  liftShowsPrec sp sl d (TestF i xs) =
+    showsBinaryWith showsPrec (const sl) "TestF" d i xs
+
+instance Fun.Arg a => Fun.Arg (TestF a)
+
+
+-- Note: size
+--
+-- I've found it quite difficult to get a good range of
+-- sizes for SeqT TestM Int using the basic tools in
+-- Gen. Preventing almost all examples being tiny seems to lead to
+-- some examples being unmanageably enormous. So I've decided to
+-- go with a "nuclear option". First, I select the exact total number
+-- of nodes in the SeqT. Then at each stage, the exact total size is
+-- chosen in advance to make sure the target is met.
```

Comment:
I shouldn't say "exact" because I'm not so sure it is. "Approximate" is plenty good enough.